hub/core/compression.py

```diff
@@ -620,3 +652,119 @@ def _read_mp3_shape(file: Union[bytes, memoryview, str]) -> Tuple[int, ...]:
     f_info = mp3_get_file_info if isinstance(file, str) else mp3_get_info
     info = f_info(file)
     return (info.num_frames, info.nchannels)
+
+
+def _strip_hub_mp4_header(buffer: bytes):
+    if buffer[: len(_HUB_MP4_HEADER)] == _HUB_MP4_HEADER:
+        return memoryview(buffer)[len(_HUB_MP4_HEADER) + 6 :]
+    return buffer
+
+
+def _decompress_video(
+    file: Union[bytes, memoryview, str],
+) -> np.ndarray:
+
+    shape = _read_video_shape(file)
+
+    command = [
+        FFMPEG_BINARY,
+        "-i",
+        "pipe:",
+        "-f",
+        "image2pipe",
+        "-pix_fmt",
+        "rgb24",
+        "-vcodec",
+        "rawvideo",
+        "-",
+    ]
+    if isinstance(file, str):
+        command[2] = file
+        pipe = sp.Popen(command, stdout=sp.PIPE, stderr=sp.PIPE, bufsize=10 ** 8)
+        raw_video = pipe.communicate()[0]
+    else:
+        file = _strip_hub_mp4_header(file)
+        pipe = sp.Popen(
+            command, stdin=sp.PIPE, stdout=sp.PIPE, stderr=sp.PIPE, bufsize=10 ** 8
+        )
+        raw_video = pipe.communicate(input=file)[0]
+    return np.frombuffer(raw_video[: int(np.prod(shape))], dtype=np.uint8).reshape(
+        shape
+    )
+
+
+def _read_video_shape(file: Union[bytes, memoryview, str]) -> Tuple[int, ...]:
+    info = _get_video_info(file)
+    if info["duration"] is None:
+        nframes = -1
+    else:
+        nframes = math.floor(info["duration"] * info["rate"])
+    return (nframes, info["height"], info["width"], 3)
+
+
+def _get_video_info(file: Union[bytes, memoryview, str]) -> dict:
+    duration = None
+    fps = None
+    command = [
+        FFPROBE_BINARY,
+        "-v",
+        "error",
+        "-select_streams",
+        "v:0",
+        "-show_entries",
+        "stream=width,height,duration,r_frame_rate",
+        "-of",
+        "default=noprint_wrappers=1",
+        "pipe:",
+    ]
+
+    if isinstance(file, str):
+        command[-1] = file
+        pipe = sp.Popen(command, stdout=sp.PIPE, stderr=sp.PIPE, bufsize=10 ** 5)
+        raw_info = pipe.stdout.read()
+        pipe.communicate()
+    else:
+        if file[: len(_HUB_MP4_HEADER)] == _HUB_MP4_HEADER:
+            mv = memoryview(file)
```

Comment:
can you move these functions into `hub/util/video.py`?

Reply:
@mccrearyd Can we do that as part of the compression refactor ticket later on?

Reply:
sure